Trouble building the project for the first time

Hi all 👋🏼,

I'm trying to build this project for the first time and running into a few issues. Super new to rust and haven't done any c since college so there may be obvious things I'm missing. If this is actually an llvm issue, I'm happy to take the conversation over there, I just don't know enough to tell.

```
macOS Catalina 10.15.2
rustc 1.41.0-nightly (c8ea4ace9 2019-12-14)
cargo 1.41.0-nightly (626f0f40e 2019-12-03)
```
(lemme know if there's other info that'd be useful for debugging and I can include it)

The directions in the readme worked great for me up until the LLVM step.

I brew installed `ninja`, `ccache`, and `cmake`. There doesn't seem to be a keg for clang, but I've got a version shipped with the OS. No idea if this is a useful version or not though: `Apple clang version 11.0.0 (clang-1100.0.33.12)`.

The first thing I ran into was that llvm-project doesn't have a Makefile, so `make llvm` doesn't work, but the llvm-project docs have build steps so I followed those and did `mcd build && cmake -G Ninja ../llvm` and then `ninja`. This seemed to do vaguely useful things and exit successfully? Are there other options I need to build it with?

So next I restarted iTerm (just in case!), and moved into the lumen folder and ran `make build`. This fails loading the `hasbrown` dependency with the following output:

```
➜  lumen git:(develop) make build
LLVM_SYS_90_PREFIX=`cd ~/.local/share/llvm/lumen && pwd` cargo build
/bin/sh: line 0: cd: /Users/marc/.local/share/llvm/lumen: No such file or directory
    Updating git repository `https://github.com/hansihe/hashbrown`
error: failed to load source for a dependency on `hashbrown`

Caused by:
  Unable to update https://github.com/hansihe/hashbrown#fc6e2652

Caused by:
  revspec 'fc6e265214edf3731944a9ae5e879990c743ccfe' not found; class=Reference (4); code=NotFound (-3)
make: *** [build] Error 101
```

It seems like the actual error is that lumen expects llvm things to be in `~/.local/share/llvm` and that apparently didn't happen.

I also tried passing that local/share directory to cmake and got:

```
➜  build git:(master) cmake -G Ninja ~/.local/share/llvm
CMake Error: The source directory "/Users/marc/.local/share/llvm" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
```

I tried moving `llvm-project` to `~/.local/share/llvm` which I didn't really expect to work, but I'm kind of out of ideas. There's still no lumen folder there, so it fails in the same way.

Any help would be appreciated! Happy to update the readme as well if this isn't just a me problem.

Thanks!

--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -51,6 +51,7 @@ task:
   windows_container:
     image: cirrusci/windowsservercore:cmake
     os_version: 2019
+    memory: 12
   env:
     CIRRUS_SHELL: powershell
     CARGO_HOME: $USERPROFILE\.cargo

--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -1,7 +1,7 @@
 task:
   name: Check Formatted
   container:
-    image: kronicdeth/lumen-development@sha256:6b71e10a3dc14f655a891a7e53c69a1024554acc195cef33860fba0cc523df83
+    image: kronicdeth/lumen-development@sha256:2d6eaba839218d6180f06a102cfb7424b02d177820ca980344174328df0c0fc3
   script: cargo fmt -- --check
 
 x86_64_task_template: &x86_64_TASK_TEMPLATE
@@ -23,7 +23,7 @@ x86_64_task_template: &x86_64_TASK_TEMPLATE
 task:
   name: Linux x86_64
   container:
-    image: kronicdeth/lumen-development@sha256:6b71e10a3dc14f655a891a7e53c69a1024554acc195cef33860fba0cc523df83
+    image: kronicdeth/lumen-development@sha256:2d6eaba839218d6180f06a102cfb7424b02d177820ca980344174328df0c0fc3
     cpu: 4
     memory: 12
   linux_x86_64_cargo_cache:
@@ -92,7 +92,7 @@ task:
 task:
   name: Linux wasm32
   container:
-    image: kronicdeth/lumen-development@sha256:6b71e10a3dc14f655a891a7e53c69a1024554acc195cef33860fba0cc523df83
+    image: kronicdeth/lumen-development@sha256:2d6eaba839218d6180f06a102cfb7424b02d177820ca980344174328df0c0fc3
     memory: 6
   linux_wasm32_cargo_cache:
     folder: $CARGO_HOME/registry

--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -137,6 +137,8 @@ task:
   install_wasm32_target_script: rustup target add wasm32-unknown-unknown --toolchain nightly
   install_wasm_bindgen_cli_script: cargo +nightly install wasm-bindgen-cli
   install_wasm_pack_script: curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
+  update_homebrew_script: |
+    brew update
   install_chrome_script: |
     brew cask install google-chrome
     brew cask info google-chrome

--- a/.cirrus/README.md
+++ b/.cirrus/README.md
@@ -1,0 +1,4 @@
+## Rebuilding
+
+1. Build the Docker image: `docker build --no-cache -t kronicdeth/lumen-development:latest .` 
+2. Push the image to Docker Hub: `docker push kronicdeth/lumen-development:latest`

--- a/examples/interpreter-in-browser/src/lib.rs
+++ b/examples/interpreter-in-browser/src/lib.rs
@@ -78,7 +78,7 @@ where
 
 pub fn lower(input: &str, config: ParseConfig) -> Result<Module, ()> {
     let (parsed, parser): (ErlAstModule, _) = parse(input, config);
-    let (res, messages) = lower_module(&parsed);
+    let (res, messages) = lower_module(&parser.config.codemap.lock().unwrap(), &parsed);
 
     let emitter = StandardStreamEmitter::new(ColorChoice::Auto).set_codemap(parser.config.codemap);
     for err in messages.iter() {

--- a/liblumen_alloc/src/blocks/free_block_tree.rs
+++ b/liblumen_alloc/src/blocks/free_block_tree.rs
@@ -54,7 +54,7 @@ impl FreeBlocks {
         let mut cursor = self.addr_tree.front();
         let mut result = None;
 
-        let aligned = layout.pad_to_align().unwrap();
+        let aligned = layout.pad_to_align();
         let requested = aligned.size();
 
         while let Some(block) = cursor.get() {
@@ -77,7 +77,7 @@ impl FreeBlocks {
         let mut result = None;
         let mut best_size = 0;
 
-        let aligned = layout.pad_to_align().unwrap();
+        let aligned = layout.pad_to_align();
         let requested = aligned.size();
 
         match self.order {

--- a/liblumen_alloc/src/erts/fragment.rs
+++ b/liblumen_alloc/src/erts/fragment.rs
@@ -123,7 +123,7 @@ impl HeapAlloc for HeapFragment {
         use liblumen_core::sys::sysconf::MIN_ALIGN;
 
         // Ensure layout has alignment padding
-        let layout = layout.align_to(MIN_ALIGN).unwrap().pad_to_align().unwrap();
+        let layout = layout.align_to(MIN_ALIGN).unwrap().pad_to_align();
         // Capture the base pointer for this allocation
         let top = self.heap_top() as *mut u8;
         // Calculate available space and fail if not enough is free

--- a/liblumen_alloc/src/erts/process/gc/old_heap.rs
+++ b/liblumen_alloc/src/erts/process/gc/old_heap.rs
@@ -85,7 +85,7 @@ impl HeapAlloc for OldHeap {
     unsafe fn alloc_layout(&mut self, layout: Layout) -> AllocResult<NonNull<Term>> {
         use liblumen_core::sys::sysconf::MIN_ALIGN;
 
-        let layout = layout.align_to(MIN_ALIGN).unwrap().pad_to_align().unwrap();
+        let layout = layout.align_to(MIN_ALIGN).unwrap().pad_to_align();
 
         let needed = layout.size();
         let available = self.heap_available() * mem::size_of::<Term>();

--- a/liblumen_alloc/src/erts/process/gc/sweep.rs
+++ b/liblumen_alloc/src/erts/process/gc/sweep.rs
@@ -72,8 +72,7 @@ where
 
         let layout = Layout::from_size_align(words * mem::size_of::<Term>(), MIN_ALIGN)
             .unwrap()
-            .pad_to_align()
-            .unwrap();
+            .pad_to_align();
         let total_size = layout.size();
 
         // Allocate space for move

--- a/liblumen_alloc/src/erts/process/gc/young_heap.rs
+++ b/liblumen_alloc/src/erts/process/gc/young_heap.rs
@@ -353,7 +353,7 @@ impl HeapAlloc for YoungHeap {
     unsafe fn alloc_layout(&mut self, layout: Layout) -> AllocResult<NonNull<Term>> {
         use liblumen_core::sys::sysconf::MIN_ALIGN;
 
-        let layout = layout.align_to(MIN_ALIGN).unwrap().pad_to_align().unwrap();
+        let layout = layout.align_to(MIN_ALIGN).unwrap().pad_to_align();
 
         let needed = layout.size();
         let available = self.heap_available() * mem::size_of::<Term>();

--- a/liblumen_alloc/src/erts/term/binary/heap.rs
+++ b/liblumen_alloc/src/erts/term/binary/heap.rs
@@ -90,7 +90,7 @@ impl HeapBin {
         // We pad to alignment so that the Layout produced here
         // matches that returned by `Layout::for_value` on the
         // final `HeapBin`
-        let layout = unpadded_layout.pad_to_align().unwrap();
+        let layout = unpadded_layout.pad_to_align();
 
         (layout, flags_offset, data_offset)
     }

--- a/liblumen_alloc/src/erts/term/binary/process.rs
+++ b/liblumen_alloc/src/erts/term/binary/process.rs
@@ -134,7 +134,7 @@ impl ProcBin {
         // We pad to alignment so that the Layout produced here
         // matches that returned by `Layout::for_value` on the
         // final `ProcBinInner`
-        let layout = unpadded_layout.pad_to_align().unwrap();
+        let layout = unpadded_layout.pad_to_align();
 
         unsafe {
             let non_null = sys_alloc::alloc(layout)?;

--- a/liblumen_alloc/src/erts/term/closure.rs
+++ b/liblumen_alloc/src/erts/term/closure.rs
@@ -59,7 +59,7 @@ impl ClosureLayout {
         let (layout, code_offset) = layout.extend(Layout::new::<Option<Code>>()).unwrap();
         let (layout, env_offset) = layout.extend(Layout::for_value(env)).unwrap();
 
-        let layout = layout.pad_to_align().unwrap();
+        let layout = layout.pad_to_align();
 
         Self {
             layout,

--- a/liblumen_alloc/src/erts/term/tuple.rs
+++ b/liblumen_alloc/src/erts/term/tuple.rs
@@ -112,7 +112,7 @@ impl Tuple {
         // We pad to alignment so that the Layout produced here
         // matches that returned by `Layout::for_value` on the
         // final `Tuple`
-        let layout = base_layout.pad_to_align().unwrap();
+        let layout = base_layout.pad_to_align();
 
         (layout, data_offset)
     }

--- a/liblumen_alloc/src/erts/testing.rs
+++ b/liblumen_alloc/src/erts/testing.rs
@@ -142,7 +142,7 @@ impl HeapAlloc for RegionHeap {
     /// process heap during garbage collection
     unsafe fn alloc_layout(&mut self, layout: Layout) -> AllocResult<NonNull<Term>> {
         // Ensure layout has alignment padding
-        let layout = layout.align_to(MIN_ALIGN).unwrap().pad_to_align().unwrap();
+        let layout = layout.align_to(MIN_ALIGN).unwrap().pad_to_align();
         // Capture the base pointer for this allocation
         let top = self.top;
         // Calculate available space and fail if not enough is free

--- a/liblumen_beam/src/beam.rs
+++ b/liblumen_beam/src/beam.rs
@@ -13,8 +13,7 @@
 //!
 //! ## References
 //!
-//! * [BEAM Wisdom - BEAM File
-//!   Format](http://beam-wisdoms.clau.se/en/latest/indepth-beam-file.html#)
+//! * [BEAM Wisdom - BEAM File Format](http://beam-wisdoms.clau.se/en/latest/indepth-beam-file.html#)
 //! * [The BEAM Book - The BEAM File Format](https://happi.github.io/theBeamBook/#BEAM_files)
 //!
 //! ## Alternative Implementations

--- a/liblumen_beam/src/beam/reader/beam_file.rs
+++ b/liblumen_beam/src/beam/reader/beam_file.rs
@@ -42,8 +42,7 @@ impl<C: Chunk> BeamFile<C> {
     /// Returns all chunks in the order they were encountered in the origin BEAM file
     ///
     /// ## Alternative Implementations
-    /// - [`org.elixir_lang.beam.Beam#chunkCollection` in IntelliJ
-    ///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+    /// - [`org.elixir_lang.beam.Beam#chunkCollection` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
     ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/Beam.kt#L70) in Kotlin.
     pub fn chunks(&self) -> Vec<&C> {
         let mut result: Vec<&C> = Vec::new();
@@ -70,8 +69,7 @@ impl<C: Chunk> BeamFile<C> {
     ///     }
     ///
     /// ## Alternative Implementations
-    /// - [`org.elixir_lang.beam.Beam#chunk` in IntelliJ
-    ///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+    /// - [`org.elixir_lang.beam.Beam#chunk` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
     ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/Beam.kt#L68-L69) in
     ///   Kotlin.
     pub fn get_chunk(&self, id: &Id) -> Option<&C> {
@@ -81,8 +79,7 @@ impl<C: Chunk> BeamFile<C> {
     /// Returns whichever chunk is the atom chunk, if it exists
     ///
     /// ## Alternative Implementations
-    /// - [`org.elixir_lang.beam.Beam#atoms` in IntelliJ
-    ///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+    /// - [`org.elixir_lang.beam.Beam#atoms` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
     ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/Beam.kt#L63-L65) in
     ///   Kotlin.
     pub fn atoms(&self) -> Option<&C> {

--- a/liblumen_beam/src/beam/reader/chunk.rs
+++ b/liblumen_beam/src/beam/reader/chunk.rs
@@ -114,8 +114,7 @@ impl Chunk for RawChunk {
 
     /// ## Alternative Implementations
     ///
-    /// - [`org.elixir_lang.beam.chunk.Chunk.from` in IntelliJ
-    ///   ELixir](https://github.com/KronicDeth/intellij-elixir/blob/
+    /// - [`org.elixir_lang.beam.chunk.Chunk.from` in IntelliJ ELixir](https://github.com/KronicDeth/intellij-elixir/blob/
     ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/Chunk.java#
     ///   L34-L52) in Kotlin
     fn decode_data<R: Read>(id: &Id, mut reader: R) -> Result<Self>
@@ -212,11 +211,9 @@ impl Chunk for AtomChunk {
 /// - [StrTChunk](StrTChunk) for strings from the string pool used in `bs_*` operations.
 ///
 /// ## Alternative Implementations
-/// - [`org.elixir_lang.beam.chunk.Code` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [`org.elixir_lang.beam.chunk.Code` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/Code.kt) in Kotlin
-/// - [`org.elixir_lang.beam.chunk.code.Operation#assembly` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [`org.elixir_lang.beam.chunk.code.Operation#assembly` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/code/Operation.kt#
 ///   L23-L164) in Kotlin for an example of how to disassemble the individual byte code operations
 ///   back to BEAM Assembly.
@@ -246,8 +243,7 @@ impl Chunk for CodeChunk {
     }
 
     /// ## Alternative Implementations
-    /// - [`org.elixir_lang.beam.chunk.Code.Companion.from` in IntelliJ
-    ///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+    /// - [`org.elixir_lang.beam.chunk.Code.Companion.from` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
     ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/Code.kt#L171-L216)
     ///   in Kotlin
     /// NOTE: This implementation decodes the operations as it loads the data instead of storing it
@@ -286,8 +282,7 @@ impl Chunk for CodeChunk {
 /// `'static str`.
 ///
 /// ## Alternative Implementations
-/// - [`org.elixir_lang.beam.chunk.Strings` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [`org.elixir_lang.beam.chunk.Strings` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/Strings.kt) in Kotlin
 #[derive(Debug, PartialEq, Eq)]
 pub struct StrTChunk {
@@ -329,8 +324,7 @@ impl Chunk for StrTChunk {
 /// operands for the MFA.
 ///
 /// ## Alternative Implementations
-/// - [`org.elixir_lang.beam.chunk.Imports` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [`org.elixir_lang.beam.chunk.Imports` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/Imports.kt) in Kotlin
 #[derive(Debug, PartialEq, Eq)]
 pub struct ImpTChunk {
@@ -384,8 +378,7 @@ impl Chunk for ImpTChunk {
 /// The format is the same as [LocTChunk](LocTChunk).
 ///
 /// ## Alternative Implementations
-/// - [`org.elixir_lang.beam.chunk.CallDefinitions` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [`org.elixir_lang.beam.chunk.CallDefinitions` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/CallDefinitions.java)
 ///   in Java
 #[derive(Debug, PartialEq, Eq)]
@@ -399,8 +392,7 @@ impl Chunk for ExpTChunk {
     }
 
     /// ## Alternative Implementations
-    /// - [`org.elixir_lang.beam.chunk.CallDefinitions.from(Chunk, Chunk.TypeID.EXPT,
-    ///   Atoms)`](https://github.com/KronicDeth/intellij-elixir/blob/
+    /// - [`org.elixir_lang.beam.chunk.CallDefinitions.from(Chunk, Chunk.TypeID.EXPT, Atoms)`](https://github.com/KronicDeth/intellij-elixir/blob/
     ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/CallDefinitions.
     ///   java#L52-L76) in Java
     fn decode_data<R: Read>(id: &Id, mut reader: R) -> Result<Self>
@@ -436,8 +428,7 @@ impl Chunk for ExpTChunk {
 /// `term_to_binary`.
 ///
 /// ## Alternative Implementations
-/// - [`org.elixir_lang.beam.chunk.Literals` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [`org.elixir_lang.beam.chunk.Literals` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/Literals.kt) in Kotlin
 #[derive(Debug, PartialEq, Eq)]
 pub struct LitTChunk {
@@ -505,8 +496,7 @@ impl Chunk for LitTChunk {
 /// The format is the same as [ExpTChunk](ExpTChunk).
 ///
 /// ## Alternative Implementations
-/// - [`org.elixir_lang.beam.chunk.CallDefinitions` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [`org.elixir_lang.beam.chunk.CallDefinitions` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/CallDefinitions.java)
 ///   in Java
 #[derive(Debug, PartialEq, Eq)]
@@ -520,8 +510,7 @@ impl Chunk for LocTChunk {
     }
 
     /// ## Alternative Implementations
-    /// - [`org.elixir_lang.beam.chunk.CallDefinitions.from(Chunk, Chunk.TypeID.LOCT,
-    ///   Atoms)`](https://github.com/KronicDeth/intellij-elixir/blob/
+    /// - [`org.elixir_lang.beam.chunk.CallDefinitions.from(Chunk, Chunk.TypeID.LOCT, Atoms)`](https://github.com/KronicDeth/intellij-elixir/blob/
     ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/CallDefinitions.
     ///   java#L52-L76) in Java
     fn decode_data<R: Read>(id: &Id, mut reader: R) -> Result<Self>
@@ -564,8 +553,7 @@ impl Chunk for LocTChunk {
 /// 6. Old Unique - a number that _was_ used to uniquely identify the anonymous function
 ///
 /// ## Alternative Implementations
-/// - [`org.elixir_lang.beam.chunk.Functions` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [`org.elixir_lang.beam.chunk.Functions` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/Functions.kt) in
 ///   Kotlin
 #[derive(Debug, PartialEq, Eq)]
@@ -736,8 +724,7 @@ impl Chunk for DbgiChunk {
     }
 
     /// ## Alternative Implementations
-    /// - [`org.elixir_lang.beam.chunk.debug_info` namespace in IntelliJ
-    ///   Elixir](https://github.com/KronicDeth/intellij-elixir/tree/
+    /// - [`org.elixir_lang.beam.chunk.debug_info` namespace in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/tree/
     ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/debug_info)
     ///   handles all current variants, both Erlang and Elixir and invalid forms thereof.
     fn decode_data<R: Read>(id: &Id, mut reader: R) -> Result<Self>

--- a/liblumen_beam/src/beam/reader/chunk/auxiliary.rs
+++ b/liblumen_beam/src/beam/reader/chunk/auxiliary.rs
@@ -16,8 +16,7 @@ impl Header {
         }
     }
     /// Alternative Implementations
-    /// - [In `org.elixir_lang.bean.chunk.Chunk.from` in IntelliJ
-    ///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+    /// - [In `org.elixir_lang.bean.chunk.Chunk.from` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
     ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/Chunk.java#
     ///   L36-L40) in Java.
     pub fn decode<R: std::io::Read>(mut reader: R) -> std::io::Result<Self> {
@@ -34,8 +33,7 @@ impl Header {
 }
 
 /// ## Alternative Implementations
-/// - [In `org.elixir_lang.bean.chunk.Chunk.from` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [In `org.elixir_lang.bean.chunk.Chunk.from` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/Chunk.java#L45) in
 ///   Java.
 pub fn padding_size(data_size: u32) -> u32 {

--- a/liblumen_beam/src/beam/reader/parts.rs
+++ b/liblumen_beam/src/beam/reader/parts.rs
@@ -23,8 +23,7 @@ pub struct Atom {
 /// An imported function.
 ///
 /// ## Alternative Implementations
-/// - [`org.elixir_lang.beam.chunk.imports.Import` in IntelliJ
-///   Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
+/// - [`org.elixir_lang.beam.chunk.imports.Import` in IntelliJ Elixir](https://github.com/KronicDeth/intellij-elixir/blob/
 ///   2f5c826040681e258e98c3e2f02b25985cd0766b/src/org/elixir_lang/beam/chunk/imports/Import.kt) in
 ///   Kotlin
 #[derive(Debug, PartialEq, Eq)]

--- a/liblumen_compiler/src/compiler.rs
+++ b/liblumen_compiler/src/compiler.rs
@@ -116,9 +116,14 @@ impl Compiler {
     // Compiles a .erl file to Erlang AST
     fn parse_erl(&self, parser: &mut Parser, file: &Path) -> Result<Module> {
         use libeir_syntax_erl::ast;
+        let codemap = parser.config.codemap.clone();
         match parser.parse_file::<&Path, ast::Module>(file) {
             Ok(ast) => {
-                let (res, messages) = libeir_syntax_erl::lower_module(&ast);
+                let (res, messages) = {
+                    // TODO: Revisit this if the same codemap is used across threads
+                    let codemap_guard = codemap.lock().unwrap();
+                    libeir_syntax_erl::lower_module(&*codemap_guard, &ast)
+                };
                 for msg in messages.iter() {
                     self.diagnostic(&msg.to_diagnostic());
                 }

--- a/liblumen_eir_interpreter/src/bin.rs
+++ b/liblumen_eir_interpreter/src/bin.rs
@@ -42,7 +42,7 @@ where
     P: AsRef<Path>,
 {
     let (parsed, parser): (ErlAstModule, _) = parse_file(path, config);
-    let (res, messages) = lower_module(&parsed);
+    let (res, messages) = lower_module(&parser.config.codemap.lock().unwrap(), &parsed);
 
     let emitter =
         StandardStreamEmitter::new(ColorChoice::Auto).set_codemap(parser.config.codemap.clone());

--- a/liblumen_eir_interpreter/src/exec/mod.rs
+++ b/liblumen_eir_interpreter/src/exec/mod.rs
@@ -313,7 +313,7 @@ impl CallExecutor {
 
         let mut exec = self;
         // Outer loop for optimized execution within the current function
-        'outer: loop {
+        loop {
             // Insert block argument into environment
             let block_arg_vals = fun.fun.block_args(block);
             //trace!("{:?} {:?}", &block_arg_vals, &exec.next_args);

--- a/liblumen_eir_interpreter/src/tests/mod.rs
+++ b/liblumen_eir_interpreter/src/tests/mod.rs
@@ -33,7 +33,7 @@ where
 
 pub fn lower(input: &str, config: ParseConfig) -> Result<Module, ()> {
     let (parsed, parser): (ErlAstModule, _) = parse(input, config);
-    let (res, messages) = lower_module(&parsed);
+    let (res, messages) = lower_module(&parser.config.codemap.lock().unwrap(), &parsed);
 
     let emitter =
         StandardStreamEmitter::new(ColorChoice::Auto).set_codemap(parser.config.codemap.clone());

